File: src/main/java/org/luxons/sevenwonders/game/moves/DiscardMove.java
package org.luxons.sevenwonders.game.moves;

import java.util.List;

import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.cards.Card;

public class DiscardMove extends Move {

    DiscardMove(int playerIndex, Card card, PlayerMove move) {
        super(playerIndex, card, move);
    }

    public boolean isValid(Table table) {
        return true;
    }

    @Override
    public void place(Table table, List<Card> discardedCards, Settings settings) {
        discardedCards.add(getCard());
    }

    @Override
    public void activate(Table table, List<Card> discardedCards, Settings settings) {
        table.giveGoldForDiscarded(getPlayerIndex(), settings.getDiscardedCardGold());
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/game/moves/DiscardMove.kt
package org.luxons.sevenwonders.game.moves

import org.luxons.sevenwonders.game.PlayerContext
import org.luxons.sevenwonders.game.Settings
import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.cards.Card

internal class DiscardMove(move: PlayerMove, card: Card, player: PlayerContext) :
    CardFromHandMove(move, card, player) {

    override fun place(discardedCards: MutableList<Card>, settings: Settings) {
        discardedCards.add(card)
    }

    override fun activate(discardedCards: List<Card>, settings: Settings) {
        playerContext.board.addGold(settings.discardedCardGold)
    }
}
